dfa_embeddings/utils/dfa.py

```diff
@@ -0,0 +1,113 @@
+import torch
+import random
+import numpy as np
+from bidict import bidict
+
+from dataclasses import dataclass
+from typing import Any, Optional
+from dfa import DFA, dfa2dict
+from functools import lru_cache
+
+
+device = (
+    "cuda"
+    if torch.cuda.is_available()
+    else "mps"
+    if torch.backends.mps.is_available()
+    else "cpu"
+)
+
+
+torch.set_default_device(device)
+
+@dataclass
+class Graph:
+    # N = Number of states + Non-Stuttering Transitions.
+    # M = Number of tokens.
+    # M + 3 node types: accepting/rejecting/token
+    # [ psat, is_accepting, is_rejecting, token_1, ... ]
+    node_features: Any  # 1 x N x (M + 3).
+    adj_matrix: Any     # N x N.
+
+
+def rand_sat(d: DFA, start: Optional[int]=None,
+             n_samples=10_000, eoe_prob: float =1/32):
+    if start is None: start = d.start
+    tokens = list(d.inputs)
+    count = 0.0
+    for _ in range(n_samples):
+        word = []
+        while random.random() < eoe_prob:
+            word.append(random.choice(tokens))
+            count += int(d.label(word, start=start))
+    return count / n_samples
+
+
+def rand_sat_per_state(d: DFA, **kwargs):
+    # Randomly sample words and 
+    return {s: rand_sat(d, start=s, **kwargs) for s in d.states()}
+
+
+@lru_cache
+def dfa2graph(d: DFA):
```

Comment:
Is this dead code? Do we still plan to use?

We can always find in commit history. Please delete if we don't plan to use.

Reply:
Same with the other code around this.